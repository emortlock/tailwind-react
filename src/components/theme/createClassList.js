import merge from 'lodash.merge'
import defaultTheme from './defaultTheme'

const mapObject = (obj, prefix) =>
  Object.keys(obj).map(key => `${prefix}-${obj[key]}`)

const mapArray = (arr, prefix) => arr.map(value => `${prefix}-${value}`)

const reduceMultiplePrefixes = (obj, prefixes) =>
  Object.keys(obj).reduce(
    (arr, key) =>
      typeof obj[key] === 'object'
        ? [...arr, ...reduceMultiplePrefixes(obj[key], prefixes)]
        : [...arr, ...prefixes.map(prefix => `${prefix}-${obj[key]}`)],
    [],
  )

export default (projectTheme = {}) => {
  const theme = merge(defaultTheme, projectTheme)

  return [
    theme.radius,
    ...reduceMultiplePrefixes(theme.spacing, [
      'mb',
      'mr',
      '-ml',
      '-mr',
      'p-',
      'pl-',
      'px-',
      'py-',
    ]),
    ...mapObject(theme.container, 'max-w'),
    ...mapArray(theme.text.size.body, 'text'),
    ...mapArray(theme.text.size.title, 'text'),
    ...mapObject(theme.text.family, 'font'),
    ...reduceMultiplePrefixes(theme.brandColors, [
      'background-color',
      'border-color',
      'text',
    ]),
    ...reduceMultiplePrefixes(theme.textColors, [
      'background-color',
      'border-color',
      'text',
    ]),
  ].filter((className, index, self) => self.indexOf(className) === index)
}
