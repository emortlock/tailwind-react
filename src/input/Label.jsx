import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'

import { withConfig } from '../config'

const Label = ({
  config,
  is,
  children,
  className,
  htmlFor,
  disabled,
  ...rest
}) => {
  const Component = is

  return (
    <Component
      {...rest}
      className={classnames(
        `mb-${config.spacing.sm}`,
        'inline-block',
        className,
        disabled && 'opacity-50',
      )}
      htmlFor={htmlFor}
    >
      {children}
    </Component>
  )
}

Label.propTypes = {
  config: PropTypes.shape({}).isRequired,
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  children: PropTypes.node,
  className: PropTypes.string,
  htmlFor: PropTypes.string.isRequired,
  disabled: PropTypes.bool,
}

Label.defaultProps = {
  is: 'label',
  children: undefined,
  className: undefined,
  disabled: false,
}

export { Label as component }
export default withConfig(Label)
