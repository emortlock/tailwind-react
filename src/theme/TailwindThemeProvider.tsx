import { Theme } from './TailwindTheme'

import React from 'react'
import merge from 'lodash.merge'

import { defaultTheme } from './defaultTheme'
import { TailwindTheme } from './TailwindTheme'

export const TailwindThemeProvider = ({
  theme = {},
  children,
}: {
  theme: Partial<Theme>
  children: React.ReactNode
}) => {
  const mergedTheme: Theme = merge({}, defaultTheme, theme)

  return (
    <TailwindTheme.Provider value={mergedTheme}>
      {children}
    </TailwindTheme.Provider>
  )
}
