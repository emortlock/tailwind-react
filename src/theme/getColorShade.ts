const shades = ['100', '200', '300', '400', '500', '600', '700', '800', '900']

export const getColorShade = (
  color?: string | string[],
  offset: string | number = 1,
): string => {
  if (!color) return ''

  let currentColor: string | string[] = Array.isArray(color) ? color[0] : color

  if (offset === 0) return currentColor

  currentColor =
    currentColor === 'white' ? ['gray', '100'] : currentColor.split('-')
  let shadeOffset = offset

  if (color === 'white') {
    if (shadeOffset < 1) return color
    if (shadeOffset === 1) return 'gray-100'

    currentColor = ['gray', '100']
    if (typeof offset === 'number') shadeOffset = offset - 1
  }

  if (currentColor.length === 1) {
    currentColor.push('400')
  }

  if (typeof shadeOffset === 'string') {
    return `${currentColor[0]}-${shadeOffset}`
  }

  const shadeIndex = Math.min(
    Math.max(shades.indexOf(currentColor[1]) + shadeOffset, 0),
    shades.length - 1,
  )
  const newShade = shades[shadeIndex]

  return `${currentColor[0]}-${newShade}`
}
