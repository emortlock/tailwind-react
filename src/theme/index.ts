export * from './defaultTheme'
export * from './TailwindTheme'
export * from './TailwindThemeProvider'
export * from './withTheme'
export * from './useThemeValue'
export * from './getColorShade'
