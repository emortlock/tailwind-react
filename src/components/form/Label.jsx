import React from 'react'
import PropTypes from 'prop-types'

import { withTheme } from '../theme'
import { BaseComponent } from '../tailwind'

const Label = ({
  theme,
  is,
  field: { inputId, disabled },
  children,
  htmlFor,
  ...rest
}) => (
  <BaseComponent
    is={is}
    inlineBlock
    htmlFor={inputId || htmlFor}
    m={{ b: theme.spacing.sm }}
    opacity={disabled ? 50 : undefined}
    {...rest}
  >
    {children}
  </BaseComponent>
)

Label.propTypes = {
  theme: PropTypes.shape({}).isRequired,
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  field: PropTypes.shape({
    inputId: PropTypes.string,
    disabled: PropTypes.bool,
  }),
  children: PropTypes.node,
  htmlFor: PropTypes.string,
}

Label.defaultProps = {
  is: 'label',
  field: {
    disabled: false,
  },
  children: undefined,
  htmlFor: undefined,
}

export { Label as component }
export default withTheme(Label)
