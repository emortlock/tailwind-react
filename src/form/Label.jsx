import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'

import { withTheme } from '../theme'
import {
  getTailwindClassNames,
  tailwindProps,
  tailwindPropTypes,
} from '../tailwind'
import { filterProps } from '../utils'

const Label = ({
  theme,
  is,
  field: { inputId, disabled },
  children,
  className,
  htmlFor,
  ...rest
}) => {
  const Component = is
  const userClassNames = classnames(getTailwindClassNames(rest), className)

  return (
    <Component
      {...filterProps(rest, [...tailwindProps, 'invalid'])}
      className={classnames(
        `mb-${theme.spacing.sm}`,
        'inline-block',
        disabled && 'opacity-50',
        userClassNames,
      )}
      htmlFor={inputId || htmlFor}
    >
      {children}
    </Component>
  )
}

Label.propTypes = {
  theme: PropTypes.shape({}).isRequired,
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  field: PropTypes.shape({
    inputId: PropTypes.string,
    disabled: PropTypes.bool,
  }),
  children: PropTypes.node,
  className: PropTypes.string,
  htmlFor: PropTypes.string,
  ...tailwindPropTypes,
}

Label.defaultProps = {
  is: 'label',
  field: {
    disabled: false,
  },
  children: undefined,
  className: undefined,
  htmlFor: undefined,
}

export { Label as component }
export default withTheme(Label)
