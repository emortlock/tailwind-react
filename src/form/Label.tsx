import type { TextProps } from 'tailwind-react-primitives'
import type { WithTheme } from '../theme'
import type { WithField } from './Field'

import React from 'react'

import { RawText as Text } from 'tailwind-react-primitives'
import { withTheme } from '../theme'

export interface LabelProps<E extends HTMLElement = HTMLLabelElement>
  extends WithTheme,
    WithField,
    TextProps<E> {
  hasHelp?: boolean
  hasError?: boolean
  disabled?: boolean
  optionList?: boolean
  htmlFor?: string
}

export const RawLabel = <E extends HTMLElement = HTMLLabelElement>({
  theme,
  as = 'label',
  id,
  field: {
    labelId = undefined,
    inputId = undefined,
    disabled = undefined,
  } = {},
  children,
  htmlFor,
  optionList,
  ...rest
}: LabelProps<E>) => (
  <Text<E>
    as={as}
    id={labelId || id}
    inlineBlock
    htmlFor={!optionList ? inputId || htmlFor : undefined}
    m={{ b: theme.spacing.sm }}
    opacity={disabled ? 50 : undefined}
    weight="bold"
    {...rest}
  >
    {children}
  </Text>
)

export const Label = withTheme<LabelProps>(RawLabel)
