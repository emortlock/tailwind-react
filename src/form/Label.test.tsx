import React from 'react'
import { render, RenderResult } from '@testing-library/react'

import { RawLabel as Label } from './Label'
import { defaultTheme } from '../theme'

const setup = (
  testProps = {},
): [RenderResult, { props: Record<string, any> }] => {
  const props = Object.assign(
    { theme: defaultTheme, htmlFor: 'input' },
    testProps,
  )

  const wrapper = render(<Label {...props}>Username</Label>)

  return [
    wrapper,
    {
      props,
    },
  ]
}

describe('Label', () => {
  it('renders matching snapshot', () => {
    const [{ asFragment }] = setup()

    expect(asFragment()).toMatchSnapshot()
  })
})
