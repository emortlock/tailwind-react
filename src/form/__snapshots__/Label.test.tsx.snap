// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Label renders matching snapshot 1`] = `
<DocumentFragment>
  <label
    class="font-bold leading-normal inline-block mb-2"
    for="input"
  >
    Username
  </label>
</DocumentFragment>
`;
